### PR TITLE
Add generic Locker/Unlocker

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

**Review by floitsch — APPROVED** [review]
LGTM!

File: src/mutex.cc [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    int previous_level = previous_locker->mutex_->level();
+    if (level <= previous_level) {
+      FATAL("trying to take lock of level %d (%s) while holding lock of level %d (%s)",
+          level, mutex_->name(), previous_level, previous_locker->mutex_->name());
+    }
+  }
+  // Lock after checking the precondition to avoid deadlocking
+  // instead of just failing the precondition check.
+  mutex_->lock();
+  // Only update variables after we have the lock - that grants right
```
Comment (kasperl) [inline src/mutex.cc:43]: This feels a little bit misguided. Both the Locker and the Thread are thread-local, unless Locker is used in a bizarre non-stack-allocated way.

Comment (floitsch) [reply src/mutex.cc:43]: I agree. We could remove the comment.

**Review by floitsch — APPROVED** [review]
Still LGTM.

**Pull request merged**